### PR TITLE
Support torch.optim.lr_scheduler.ReduceLROnPlateau

### PR DESCRIPTION
**Pull request by vikmary**
Support of torch.optim.lr_scheduler.ReduceLROnPlateau by implementing pytorch_lightning.Callback .

Fixes https://github.com/williamFalcon/pytorch-lightning/issues/298.

ReduceLROnPlateau should be inited as any other scheduler:
```python
class MyModule(pl.LightningModule):
  def configure_optimizers(self):
    optimizer = optim.Adam(self.parameters(), lr=self.params.learning_rate)
    scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer,
                                                     mode='min',
                                                     factor=0.1,
                                                     patience=10,
                                                     min_lr=1e-6,
                                                     verbose=True)
    return [optimizer], [scheduler]
```
From contribution guide:
- [UI](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md#no-pytorch-interference) condition met
- [simplicity of internal code](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md#simple-internal-code) is placed under review.
- seems like no doc updates needed.

### CODE REVIEW

**Comment by williamFalcon** [comment]
@vikmary looks like some of the old changes leaked into this PR. Could you rebase master into this and submit again? Thanks!

**Comment by vikmary** [comment]
> @vikmary looks like some of the old changes leaked into this PR. Could you rebase master into this and submit again? Thanks!

Merged origin master again.

**Comment by williamFalcon** [comment]
ummm. still pulls in the wrong master. 

can you pull your local master first then apply the rebase to your branch?  

basically all the messaging and ddp stuff shouldn’t change

**Comment by vikmary** [comment]
> ummm. still pulls in the wrong master.
> 
> can you pull your local master first then apply the rebase to your branch?
> 
> basically all the messaging and ddp stuff shouldn’t change

That's what I did. Can you detect what's wrong?

```bash
cd my-fork
git checkout master
git pull https://github.com/williamFalcon/pytorch-lightning.git master
git checkout support-reduceonplateau-lr-scheduler
git merge master
git push
```

**Comment by vikmary** [comment]
> ummm. still pulls in the wrong master.
> 
> can you pull your local master first then apply the rebase to your branch?
> 
> basically all the messaging and ddp stuff shouldn’t change

Diff looks strange. That's i guess what you are talking about. I'll redo the merging, thanks

**Comment by williamFalcon** [comment]
@vikmary ah it looks great now haha. I added a comment in the review

**Comment by williamFalcon** [comment]
ummm. ok. 

My question was about why we need a custom implementation of the reduceLROnPlateu?  We should be using the default PyTorch one

**Comment by williamFalcon** [comment]
(if you click on the "Files Changed" tab you'll see my original comment and links

**Comment by vikmary** [comment]
Sorry, my "Files Changes" tab is crystal clear from comments. Did you click "Add single comment" when submitting a comment or "Start a Review"? If you start a review, then you have to submit the whole review before in order for comments to appear (as it seems to me).

We ARE using the default PyTorch implementation of ReduceLROnPlateau. We even can get rid of  my class implementation (it only makes code cleaner).

The only problem with default PyTorch implementation it that we have to pass VALIDATION LOSS ato `scheduler.step` method. That is exactly what `pytorch-lightning/callbacks:ReduceLROnPlateau` class is doing.

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -768,12 +769,25 @@ def init_optimizers(self, optimizers):
...
 
         # single list or tuple
         elif isinstance(optimizers, list) or isinstance(optimizers, tuple):
             return optimizers, []
 
+    def configure_schedulers(self, schedulers):
+        custom_schedulers = []
+        i = 0
+        while i < len(schedulers):
+            if isinstance(schedulers[i], torch.optim.lr_scheduler.ReduceLROnPlateau):
```
Comment (vikmary) [inline pytorch_lightning/trainer/trainer.py]: proof here

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 2]
```diff
@@ -144,6 +144,36 @@ def on_train_end(self, logs=None):
...
+        if current is None:
+            print('ReduceLROnPlateau conditioned on metric `%s` '
+                  'which is not available. Available metrics are: %s' %
+                  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning)
+            exit(-1)
+
+        for scheduler in self.schedulers:
+            scheduler.step(current, epoch=epoch)
+
+
```
[the remark this answers is not in the document]
Comment (vikmary) [reply pytorch_lightning/callbacks/pt_callbacks.py]: `ReduceLROnPlateauScheduler.schedulers` is a list of orginal `torch.optim.lr_scheduler.ReduceLROnPlateau`, see the proof in a comment below

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -793,12 +794,25 @@ def init_optimizers(self, optimizers):
...
+                custom_schedulers.append(schedulers.pop(i))
+            i += 1
```
Comment (VismantasD) [inline pytorch_lightning/trainer/trainer.py]: There is a small issue with this snippet. When the ReduceLROnPlateau optimizer is pop'ed, the i should not be increased, otherwise, the element following the element being pop'ed ends up in position schedulers[i] and then i is immediately increased, so that element never gets checked ( to see if it is another ReduceLROnPlateau.
This would not cause any problems when there is only on ReduceLROnPlateau scheduler.
While there is probably no reason to have more than one ReduceLROnPlateau scheduler, it would be nicer to change the code. Either move the i+=1 to an else branch, or ( if we make assumption that only one ReduceLROnPlateau scheduler is present ) break out of the loop.
```suggestion
            else:
                i += 1
```

Comment (vikmary) [reply pytorch_lightning/trainer/trainer.py]: Hi, thank you for the fix. I decided to support only on ReduceLROnPlateau scheduler.

File: pytorch_lightning/callbacks/__init__.py [hunk 4]
```diff
@@ -1,7 +1,8 @@
-from .pt_callbacks import EarlyStopping, ModelCheckpoint, GradientAccumulationScheduler
+from .pt_callbacks import EarlyStopping, ModelCheckpoint, ReduceLROnPlateauScheduler, GradientAccumulationScheduler
```
Comment (Borda) [inline pytorch_lightning/callbacks/__init__.py]: it looks like a relative import which we shall not use... :)

Comment (vikmary) [reply pytorch_lightning/callbacks/__init__.py]: Relative imports of `EarlyStopping`, `ModelCheckpoint` etc. are taken from the original repository. Why is relative import of `ReduceLROnPlateauScheduler` inappropriate?

Comment (Borda) [reply pytorch_lightning/callbacks/__init__.py]: https://github.com/williamFalcon/pytorch-lightning/pull/402#issuecomment-544704568

Comment (VismantasD) [reply pytorch_lightning/callbacks/__init__.py]: #402 never fixed relative imports in callbacks __init__ as well as in many other places. I'd say that above comment is out of scope for this PR. @borda it might be better to create a separate PR that will properly fix relative imports.

Comment (Borda) [reply pytorch_lightning/callbacks/__init__.py]: It was not that the PR was fixing relative imports, but I tried to make them which was stopped...

Comment (williamFalcon) [reply pytorch_lightning/callbacks/__init__.py]: let's do a separate PR for relative imports.

**Comment by williamFalcon** [comment]
@vikmary let's get this into the next release on nov 6. can you rebase master onto this?

**Comment by vikmary** [comment]
> @vikmary let's get this into the next release on nov 6.
> 
> 1. can you rebase master onto this?
> 2. do we need that class you made? can't we just use the default pytorch one? maybe i'm missing something or reading this too quickly.
> 
> @Borda thoughts?

I'll try to rebase during today.

**Comment by Borda** [comment]
there are 3 options and I would try them in the following order:
- rebase it on `upstream/master` (upstream is the "williamFalcon") and resolve conflicts, see https://gist.github.com/ravibhure/a7e0918ff4937c9ea1c456698dcd58aa
- merge `upstream/master` and resolve conflicts
- create a clear branch from `upstream/master` and cherry-pick your commits

(for all three options maybe be easier to squash all your commits to one)

File: pytorch_lightning/callbacks/__init__.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/callbacks/__init__.py]: I have opened ticket https://github.com/williamFalcon/pytorch-lightning/issues/459

**Comment by vikmary** [comment]
The merging needs some refactoring, will not be able to finish today

**Comment by vikmary** [comment]
Hi, I refactored so that:
- there is no Callback class, we rely solely on `torch.optim.lr_scheduler.ReduceLROnPlateau`
- fixed logging error in `ModelCheckoint.on_epoch_end`

**Comment by williamFalcon** [comment]
@Borda @Ir1d merging this, look  good?

File: pytorch_lightning/trainer/train_loop_mixin.py [hunk 5]
```diff
@@ -213,7 +212,15 @@ def train(self):
...
                 for lr_scheduler in self.lr_schedulers:
-                    lr_scheduler.step(self.current_epoch)
+                    lr_scheduler.step(epoch=self.current_epoch)
+            if self.reduce_lr_on_plateau_scheduler is not None:
+                val_loss = self.callback_metrics.get('val_loss')
+                if val_loss is None:
+                    print('ReduceLROnPlateau conditioned on metric `%s` '
+                          'which is not available. Available metrics are: %s' %
+                          ('val_loss', ','.join(list(self.callback_metrics.keys()))), RuntimeWarning)
+                    exit(-1)
```
Comment (Borda) [inline pytorch_lightning/trainer/train_loop_mixin.py]: I would rather raise exception then strong `exit`

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -376,12 +378,20 @@ def init_optimizers(self, optimizers):
...
             optimizers, lr_schedulers = optimizers
+            lr_schedulers, self.reduce_lr_on_plateau_scheduler = self.configure_schedulers(lr_schedulers)
             return optimizers, lr_schedulers
 
         # single list or tuple
         elif isinstance(optimizers, list) or isinstance(optimizers, tuple):
             return optimizers, []
 
+    def configure_schedulers(self, schedulers):
+        for i in range(len(schedulers)):
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: consider `for i, scheduler for enumerate(schedulers):`

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py]: @Borda maybe make the changes for the author to get this PR in quickly?

Comment (Ir1d) [reply pytorch_lightning/trainer/trainer.py]: @williamFalcon seems we dont have permissions to edit directly..

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 7]
```diff
@@ -315,6 +315,7 @@ def on_epoch_end(self, epoch, logs=None):
                                 f'\nEpoch {epoch:05d}: {self.monitor} reached',
                                 f'{current:0.5f} (best {self.best:0.5f}), saving model to',
                                 f'{filepath} as top {self.save_top_k}')
+                        self.best = current
```
Comment (Ir1d) [inline pytorch_lightning/callbacks/pt_callbacks.py]: I think there's something wrong here. Perhaps this line should be removed?

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: correct, only maintainers and owners can do these changes (also if the PR authors click allow changes, which is here true by default)

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 7 above)
Comment (williamFalcon) [reply pytorch_lightning/callbacks/pt_callbacks.py]: done

File: pytorch_lightning/trainer/train_loop_mixin.py [hunk 8]
```diff
@@ -213,7 +214,15 @@ def train(self):
             # update LR schedulers
             if self.lr_schedulers is not None:
                 for lr_scheduler in self.lr_schedulers:
-                    lr_scheduler.step(self.current_epoch)
+                    lr_scheduler.step(epoch=self.current_epoch)
+            if self.reduce_lr_on_plateau_scheduler is not None:
+                val_loss = self.callback_metrics.get('val_loss')
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/train_loop_mixin.py]: this seems very specific. does it only need to work with val_loss?

**Comment by Ir1d** [comment]
I have one offtopic question..
Should this be epoch instead of epoch + 1?

https://github.com/williamFalcon/pytorch-lightning/pull/320/files#diff-f1ccb073775f3b2f9c294bd887086da3L202

File: pytorch_lightning/trainer/train_loop_mixin.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/train_loop_mixin.py]: I think it could be any validation metric in theory, but how do we let the user pass it in? Perhaps via a dedicated dict entry in the [validation_end](https://williamfalcon.github.io/pytorch-lightning/LightningModule/RequiredTrainerInterface/#validation_end) output similar to "log" and "progress_bar"? It's one more thing the user needs to remember, but maybe its fine since this lr_scheduler is optional and it should be mentioned in the docs.

Comment (williamFalcon) [reply pytorch_lightning/trainer/train_loop_mixin.py]: yeah, good option. let’s do it in a separate PR?

**Comment by vikmary** [comment]
Thank you
It was a long-awaiting pull-request
Hope it has some value for framework in spite of non-trivial master integration

**Comment by williamFalcon** [comment]
haha. Thank you so much for contributing! feel free to keep helping out as we’re always growing our core team!
